### DISCUSSION TITLE
Attach custom data to a Paper?

### DISCUSSION THREAD

**Discussion by GeorgiyZhuravlev** [clientIO/joint#2347] [Q&A] [answered]
### Introduction

I have implemented a custom shape to show a user on a graph:

<img width="1017" alt="image" src="https://github.com/clientIO/joint/assets/14068189/cb287f8b-4515-41f0-a116-4389cba3711e">

---

This shape is derived from `shapes.standard.Circle` with the additional `image` tag, defined in a separate file.
So, when the `userId` or other attr of a shape is updated, the setter function is called - works great!

I have an internal (component-tight, Angular) lookup object to exchange `userId` to a URL.
What I'm not quite sure about is how to better get this lookup object from within the shape setter function.
I don't feel right to do it through the global scope (window).
That's why I thought, maybe, there's a way to attach some custom data to a paper - I have access to it via `this` in a setter.
Seems, in this separate "shape definition file", only `this` is my connection point to the outer world :)
Or how would you do this, if not through globally accessible data.

```js
const staticProperties = {
  attributes: {
    userId: {
      set: function(userId, bbox, node, attrs) {
        ...
        // get image URL somehow, having `this` at this point only
        return {
          ...
          'xlink:href': DEFAULT_AVATAR, // get by `userId`
        };
      }
    },
  }
};

export const User = shapes.standard.Circle.define(
  'pcf.User',
  defaultInstanceProperties,
  prototypeProperties,
  staticProperties,
);

```

### Does your question relate to JointJS or JointJS+. Select both if applicable.

JointJS

**Comment by kumilingus** [comment, marked as answer]
Also, the fifth parameter is the cell view (note, that this is all subject to change as custom attributes are not officially out yet).

You can attach your custom data to the graph.

```js
graph.set('urls', {
  user1: './user1.png',
  user2: './user2.png',
});
```
Attribute:
```js
const staticProperties = {
  attributes: {
    userId: {
      set: function(userId, bbox, node, attrs, cellView) {
        const graph = cellView.paper.model;
        const url = graph.get('urls')[userId] || DEFAULT_IMAGE;
        return {
          'xlink:href': url
        };
      }
    },
  }
};
```

**Reply by GeorgiyZhuravlev** [reply]
Awesome! `graph.set()` saved the day :)
Thank you!

**Reply by kumilingus** [reply]
Note that the graph data (e.g. `urls`) are now part of the graph export JSON. You might want to delete it from the JSON if you use the export.

**Reply by GeorgiyZhuravlev** [reply]
Sure, I'll remove them during the "save" operation, no problem!
